qualtran/surface_code/rotation_cost_model.py

```diff
@@ -0,0 +1,109 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import abc
+import math
+
+from attrs import frozen
+
+from qualtran.surface_code.algorithm_summary import AlgorithmSummary
+
+
+class RotationCostModel(abc.ABC):
+    """Analytical estimate of the complexity of approximating a rotation given an error budget."""
+
+    @abc.abstractmethod
+    def rotation_cost(self, error_budget: float) -> AlgorithmSummary:
+        """Cost of a single rotation."""
+
+    @abc.abstractmethod
+    def prepartion_overhead(self, error_budget) -> AlgorithmSummary:
+        """Cost of preparation circuit."""
+
+
+@frozen
+class LogarithmicModel(RotationCostModel):
+    r"""A linear model in the log of the error budget with no preparation cost.
+
+    $$
+    \#T = -\textit{slope} \log_2{\textit{budget}} + \textit{overhead}
+    $$
+
+    Attributes:
+        slope: The coefficient of $log_2{budget}$.
+        overhead: The overhead.
+        gateset: A human-readable description of the gate set (e.g. 'Clifford+T').
+        approximation_protocol: A description or reference to the approximation protocol (e.g. `Diagonal:https://arxiv.org/abs/1403.2975`)
+        reference: A human-readable description of the source of the model (e.g. 'https://arxiv.org/abs/1404.5320')
+    """
+    slope: float
+    overhead: float
+    gateset: str | None = None
+    approximation_protocol: str | None = None
+    reference: str | None = None
+
+    def rotation_cost(self, error_budget: float) -> AlgorithmSummary:
+        return AlgorithmSummary(
+            t_gates=math.ceil(-self.slope * math.log2(error_budget) + self.overhead)
+        )
+
+    def prepartion_overhead(self, error_budget) -> AlgorithmSummary:
+        return AlgorithmSummary()
+
+
+@frozen
+class ConstantCostWithPreparation(RotationCostModel):
```

Comment:
again, in actual use the fact that this is for rotations is lost; but we don't want it to be _too_ long.

`ConstOverheadRotationCost` or something like that

Reply:
ptal